cstrml/market/src/lib.rs

```diff
@@ -559,74 +597,34 @@ impl<T: Trait> Module<T> {
-                    punishment.failed += 1;
-                } else {
-        
-                }
-                if punishment.success + punishment.failed >= punish_duration && punishment.value < so.amount {
-                    // 2. Do slash
-                    let real_punish_value = Self::punish_merchant(&so, &punishment, &punish_duration);
-                    punishment.value += real_punish_value;
-                    // 3. Reset Merchant Punishment
-                    punishment.success = 0;
-                    punishment.failed = 0;
-                } else {
-        
-                }
-                <MerchantPunishments<T>>::insert(&order_id, punishment.clone());
-                if punishment.value >= so.amount {
-                    Self::close_sorder(&order_id);
-                }
-            }
-        }   
-    }
-
-    pub fn punish_merchant(
-        so: &StorageOrder<T::AccountId, BalanceOf<T>>,
-        punishment: &MerchantPunishment<BalanceOf<T>>,
-        punish_duration: &EraIndex
-    ) -> BalanceOf<T> {
-        // Calculate real punish value
-        let mut real_punish_value = Zero::zero();
-        if punishment.success >= ((*punish_duration as f64) * 0.95) as EraIndex {
-        } 
-        else if punishment.success >= ((*punish_duration as f64) * 0.90) as EraIndex {
-            real_punish_value = (Perbill::from_percent(50)) * so.amount;
-        } 
-        else {
-            real_punish_value = so.amount;
+    pub fn update_merchant_punishment(order_id: &T::Hash, current_block: &BlockNumber, so_status: &OrderStatus) {
+        if let Some(mut p) = Self::merchant_punishments(order_id) {
+            match so_status {
+                OrderStatus::Success => p.success += current_block - p.updated_at,
+                OrderStatus::Failed => p.failed += current_block - p.updated_at,
+                OrderStatus::Pending => {}
+            };
+            p.updated_at = *current_block;
+            <MerchantPunishments<T>>::insert(order_id, p);
         }
+    }
 
-        // Do slash
-        if !real_punish_value.is_zero() {
-            real_punish_value = real_punish_value.min(so.amount - punishment.value);
-            T::Currency::slash(&so.merchant, real_punish_value);
-            // Update ledger
-            let mut pledge = Self::pledges(&so.merchant);
-            pledge.total -= real_punish_value;
-            pledge.used -= real_punish_value;
-            Self::upsert_pledge(&so.merchant, &pledge);
-        }
-        real_punish_value
+    // Slash pledge value for a merchant
+    pub fn slash_pledge(
+        merchant: &T::AccountId,
+        slash_value: BalanceOf<T>
+    ) {
+        T::Currency::slash(merchant, slash_value);
+        // Update ledger
+        let mut pledge = Self::pledges(merchant);
+        pledge.total -= slash_value;
+        pledge.used -= slash_value;
+        Self::upsert_pledge(merchant, &pledge);
+        //TODO: Move slash value into treasury
```

Comment:
Recording to the pm spec,  the specific `TODO` should be `50%` to user, `50%` to treasury

Reply:
We can discuss about this one. According to the doc, it seems like just 50% of pledge currency would be slashed into treasury. The other 50% pledge would be just free and used for incoming storage orders.